docs/source/ttnn/ttnn/adding_new_ttnn_operation.rst

```diff
@@ -7,174 +7,98 @@ Adding New ttnn Operation
-    }
-
-    operation::ProgramWithCallbacks create_program(const std::vector<Tensor>& input_tensors, std::vector<Tensor> &output_tensors) const {
-        Program program = ...;
-        return operation::ProgramWithCallbacks{program};
-    }
-
-    };
+C++ Implementation
+------------------
 
-    Tensor <new_operation>(const Tensor &input_tensor, bool some_arg) {
-        std::vector<Tensor> input_tensors = {input_tensor};
-        std::vector<Tensor> output_tensors operation::run(DeviceOperation(<NewOperation>{some_arg}, {input_tensor}));
-        return output_tensors[0];
-    }
+Step 1: Implement device operation (Optional)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    }  // namespace tt_metal
-    }  // namespace tt
+In order to add a new device operation, follow the directory structure shown below:
 
+`ttnn/cpp/ttnn/operations/<category>/<operation_name>/device/<operation_name>_device_operation.hpp`
+`ttnn/cpp/ttnn/operations/<category>/<operation_name>/device/<operation_name>_device_operation.cpp`
+`ttnn/cpp/ttnn/operations/<category>/<operation_name>/device/<program_factory_0>_program_factory.cpp`
 
-Add pybindings
---------------
+.. note::
+ Add as many program factories as needed
 
-In `tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp`, add the following lines
+A concrete example of a device operation can be found in `ttnn/cpp/ttnn/operations/examples/example/device`
 
-.. code-block:: cpp
+`ttnn/cpp/ttnn/operations/examples/example/device/example_device_operation.hpp`:
 
-    m_tensor.def("<new_operation>", &<new_operation>, py::arg("input_tensor").noconvert(), py::arg("some_arg").noconvert(), R"doc(
-        <NewOperation> runs new operation on input tensor.
+.. literalinclude::  examples/example/device/example_device_operation.hpp
 
-        .. csv-table::
-            :header: "Argument", "Description", "Data type", "Valid range", "Required"
+`ttnn/cpp/ttnn/operations/examples/example/device/example_device_operation.cpp`:
 
-            "input_tensor", "Input tensor", "Tensor", "Tensor of shape [W0, Z0, Y0, X0]", "Yes"
-            "some_arg", "Some arg", "bool", "Some arg to do some stuff in new operation", "Yes"
-    )doc");
+.. literalinclude::  examples/example/device/example_device_operation.cpp
 
+`ttnn/cpp/ttnn/operations/examples/example/device/single_core_program_factory.cpp`:
 
+.. literalinclude::  examples/example/device/single_core_program_factory.cpp
 
-Adding a unit test
-------------------
+`ttnn/cpp/ttnn/operations/examples/example/device/multi_core_program_factory.cpp`:
 
-Add `tests/ttnn/unit_tests/ttl/test_<new_operation>.py`:
+.. literalinclude::  examples/example/device/multi_core_program_factory.cpp
 
-.. code-block:: python
 
-    import pytest
-    import torch
-    import ttnn
+Step 2: Implement the operation in C++
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    from tests.ttnn.utils_for_testing import assert_with_pcc
+In order to add a new operation, add the following file:
```

Comment:
Lets add a brief description of this file.
It basically provides a structure and registers the operation.
The cool thing about the structure is that it allows to provide multiple overloads.
The cool thing about registering that it allows to instrument calls to an operation with goodies like "...list...".